### PR TITLE
Port uproxy-networking to freedom 0.6

### PR DESCRIPTION
**Pull request by bemasc**

### CODE REVIEW

File: src/socks-to-rtc/socks-to-rtc.ts [hunk 1]
```diff
@@ -49,7 +48,7 @@ module SocksToRtc {
 
     // Message handler queues to/from the peer.
     public signalsForPeer :Handler.Queue<WebRtc.SignallingMessage, void> =
-        new Handler.Queue<WebRtc.SignallingMessage,void>();
+        null;
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: So, there's no real need to initialise it to `null`. I would just declare it here.

File: src/socks-to-rtc/socks-to-rtc.ts [hunk 2]
```diff
@@ -73,7 +72,7 @@ module SocksToRtc {
 
     // The connection to the peer that is acting as the endpoint for the proxy
     // connection.
-    private peerConnection_  :freedom_UproxyPeerConnection.Pc = null;
+    private peerConnection_  :WebRtc.PeerConnection = null;
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: Not really related to this pull request but in the spirit of line 51 you could just declare this here.

File: src/socks-to-rtc/socks-to-rtc.ts [hunk 3]
```diff
@@ -152,16 +150,10 @@ module SocksToRtc {
...
-      if (!this.tcpServer_.isShutdown()) {
-        shutdownPromises.push(this.tcpServer_.shutdown());
-      }
-      // uproxypeerconnection doesn't allow us query whether the
-      // peerconnection has shut down but the call is explicitly idempodent.
-      shutdownPromises.push(this.peerConnection_.close());
-      return Promise.all(shutdownPromises).then((answers:any[]) => {
-        return Promise.resolve<void>();
-      });
+      // PeerConnection.close() returns void, implying that the shutdown is
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts:152]: So, what happens if `peerconnection.close()` raises an exception?

What do you think of something like:

``` typescript
return Promise.all([
  new Promise((F,R) => { this.peerConnection_.close(); F(); }),
  this.tcpServer_.shutdown
]);
```

I think this way any exceptions thrown will be returned in the "promises way".

File: src/socks-to-rtc/socks-to-rtc.ts [hunk 4]
```diff
@@ -426,6 +380,7 @@ module SocksToRtc {
         }
         log.debug(this.longId() + ': dataFromPeer: ' + data.buffer.byteLength +
             ' bytes.');
+          this.bytesReceivedFromPeer_.handle(data.buffer.byteLength);
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: indentation here

File: src/socks-to-rtc/socks-to-rtc.ts [hunk 5]
```diff
@@ -321,22 +288,10 @@ module SocksToRtc {
...
-      shutdownPromises.push(
-          this.peerConnection_.closeDataChannel(this.channelLabel_));
-      return Promise.all(shutdownPromises).then((answers:any[]) => {
-        return Promise.resolve<void>();
-      });
-    }
-
-    public handleDataFromPeer = (data:WebRtc.Data) : void => {
-      this.dataFromPeer_.handle(data);
+      // DataChannel.close() returns void, implying that it is
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts:292]: And I guess the same comment here w.r.t. propagating exceptions via promises.

File: src/socks-to-rtc/socks-to-rtc.spec.ts [hunk 6]
```diff
@@ -102,9 +104,11 @@ describe("SOCKS session", function() {
   var session :SocksToRtc.Session;
 
   var mockTcpConnection :Tcp.Connection;
-  var mockPeerconnection :freedom_UproxyPeerConnection.Pc;
+  var mockDataChannel :WebRtc.DataChannel;
   var mockBytesSent :Handler.Queue<number,void>;
+  var mockBytesReceived :Handler.Queue<number,void>;
 
+
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.spec.ts:110]: can remove this extra line

File: src/rtc-to-net/rtc-to-net.ts [hunk 7]
```diff
@@ -155,12 +154,9 @@ module RtcToNet {
...
     // Shuts down the peerconnection, fulfilling it has terminated.
-    // Since its close() method should ever reject, this should never reject.
-    // TODO: close all sessions before fulfilling
-    public stopResources_ = () : Promise<void> => {
-      // uproxypeerconnection doesn't allow us query whether the
-      // peerconnection has shut down but the call is explicitly idempodent.
-      return this.peerConnection_.close();
+    // TODO: close all sessions and return asynchronously
+    public stopResources_ = () => {
+      this.peerConnection_.close();
```
Comment (trevj) [inline src/rtc-to-net/rtc-to-net.ts]: So, like I suggested in SocksToRtc, perhaps have this return a promise? (there's an issue open to have these `stop` methods also close their sessions before fulfilling so this won't be wasted work)

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 1 above)
Comment (bemasc) [inline src/socks-to-rtc/socks-to-rtc.ts]: Done.

File: src/socks-to-rtc/socks-to-rtc.ts [hunk 8]
```diff
@@ -244,8 +218,9 @@ module SocksToRtc {
   export class Session {
     private channelLabel_ :string;
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: It looks like we don't need `channelLabel` anymore.

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 2 above)
Comment (bemasc) [inline src/socks-to-rtc/socks-to-rtc.ts]: Done.

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 3 above)
Comment (bemasc) [inline src/socks-to-rtc/socks-to-rtc.ts:152]: Done (with adaptations to make the type system happy).

File: src/rtc-to-net/rtc-to-net.ts [hunk 9]
```diff
@@ -229,6 +202,8 @@ module RtcToNet {
     // low-level WebRtc provider), then refer to dataChannel.isClosed directly.
     private isClosed_ :boolean;
 
+    private channelLabel_ :string;
```
Comment (trevj) [inline src/rtc-to-net/rtc-to-net.ts]: It doesn't look like this is needed?

File: src/churn/churn.ts [hunk 10]
```diff
@@ -12,7 +12,55 @@ var regex2dfa :any;
...
+  var log :Logging.Log = new Logging.Log('churn');
+
+  export interface ChurnSignallingMessage extends WebRtc.SignallingMessage {
+    type          :WebRtc.SignalType
+    candidate     ?:freedom_RTCPeerConnection.RTCIceCandidate;
+    description   ?:freedom_RTCPeerConnection.RTCSessionDescription;
+    churnStage :number;
+  }
+
+  export var filterCandidatesFromSdp = (sdp:string) : string => {
```
Comment (trevj) [inline src/churn/churn.ts:24]: Do we need the `var`? I think TypeScript will do the right thing without it.

(is this so you omit the fully-qualified references later in this file?)

File: src/churn/churn.ts [hunk 11]
```diff
@@ -217,16 +272,20 @@ module Churn {
...
+        this.obfuscatedConnection_.peerOpenedChannelQueue.setSyncHandler(
+            this.peerOpenedChannelQueue.handle);
         this.churnSetup_();
       });
+      // NOTE: Replacing |this.dataChannels| in this way breaks recursive nesting.
+      // If the caller or |obfuscatedConnection_| applies the same approach,
+      // the code will break in hard-to-debug fashion.  This could be
+      // addressed by using a javascript "getter", or by changing the
+      // WebRtc.PeerConnection API.
+      this.dataChannels = this.obfuscatedConnection_.dataChannels;
```
Comment (trevj) [inline src/churn/churn.ts:288]: Oooooh, public fields considered harmful. Let's file an issue to remove these from peerconnection!

File: src/churn/churn.ts [hunk 12]
```diff
@@ -244,138 +303,72 @@ module Churn {
...
-    public send = (channelLabel:string, data:WebRtc.Data) : Promise<void> => {
-      return this.obfuscatedConnection_.send(channelLabel, data);
-    }
-
-    public close = () : Promise<void> => {
-      return this.obfuscatedConnection_.close();
-    }
-
-    public onceConnected = () : Promise<WebRtc.ConnectionAddresses> => {
+    public onceConnected = new Promise<WebRtc.ConnectionAddresses>((F, R) => {
```
Comment (trevj) [inline src/churn/churn.ts]: I don't think you need to create a new promise. I think the following will work:

``` typescript
// obfuscatedConnection_ doesn't exist until onceChurnSetup_ fulfills.
public onceConnected = this.onceChurnSetup_.then(() => {
  return this.obfuscatedConnection_.onceConnected.then((addresses:WebRtc.ConnectionAddresses) => {
    this.pcState = WebRtc.State.CONNECTED;
    return addresses;
  });
});
```

File: src/churn/churn.ts [hunk 13]
```diff
@@ -244,138 +303,72 @@ module Churn {
...
+          F(addresses);
+        }, R);
       });
-    }
+    });
 
-    public onceConnecting = () : Promise<void> => {
-      return this.surrogateConnection_.onceConnecting();
-    }
+    public onceConnecting = new Promise<void>((F, R) => {
```
Comment (trevj) [inline src/churn/churn.ts]: Likewise here, I don't think you need to create a new promise.

File: src/churn/churn.ts [hunk 14]
```diff
@@ -244,138 +303,72 @@ module Churn {
...
+      this.surrogateConnection_.onceConnecting.then(() => {
+        this.pcState = WebRtc.State.CONNECTING;
+        F();
+      }, R);
+    });
 
-    public onceDisconnected = () : Promise<void> => {
+    public onceDisconnected = new Promise<void>((F, R) => {
```
Comment (trevj) [inline src/churn/churn.ts]: Likewise :-)

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 3 above)
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts:152]: :+1:

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 1 above)
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: :+1:

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 2 above)
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: :+1:

File: src/churn/churn.ts (hunk 10 above)
Comment (bemasc) [inline src/churn/churn.ts:24]: Yep, we need "var".  It doesn't compile if I remove it.

File: src/churn/churn.ts (hunk 11 above)
Comment (bemasc) [inline src/churn/churn.ts:288]: Done, #583

File: src/churn/churn.ts (hunk 10 above)
Comment (trevj) [inline src/churn/churn.ts:24]: Ah...interesting. I just played around with this in the TypeScript playground. I _think_ a plain-old JavaScript function declaration would be equivalent, e.g.:

``` typescript
module Churn
  function filterCandidatesFromSdp (sdp:string) : string {
    return sdp.split('\n').filter((s) => {
      return s.indexOf('a=candidate') != 0;
    }).join('\n');
  }
  // ...
  // ...
}
```

So whichever you prefer.

Comment (trevj) [inline src/churn/churn.ts:24]: :+1:

File: src/churn/churn.ts (hunk 11 above)
Comment (trevj) [inline src/churn/churn.ts:288]: :+1:

**Comment by trevj** [comment]
Great work. A few pretty minor things I think will help then we'll be good to merge.

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 4 above)
Comment (bemasc) [inline src/socks-to-rtc/socks-to-rtc.ts]: Done.

File: src/socks-to-rtc/socks-to-rtc.spec.ts (hunk 6 above)
Comment (bemasc) [inline src/socks-to-rtc/socks-to-rtc.spec.ts:110]: Done.

File: src/rtc-to-net/rtc-to-net.ts (hunk 7 above)
Comment (bemasc) [inline src/rtc-to-net/rtc-to-net.ts]: Done.

File: src/churn/churn.ts (hunk 10 above)
Comment (bemasc) [inline src/churn/churn.ts:24]: I suppose I'll leave this as is.  I'm a little concerned about the scoping rules if I use "function".

File: src/churn/churn.ts (hunk 12 above)
Comment (bemasc) [inline src/churn/churn.ts]: Done ... and I even de-nested it for extra Promise-ness.

File: src/churn/churn.ts (hunk 13 above)
Comment (bemasc) [inline src/churn/churn.ts]: Done.

File: src/churn/churn.ts (hunk 14 above)
Comment (bemasc) [inline src/churn/churn.ts]: Done.

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 5 above)
Comment (bemasc) [inline src/socks-to-rtc/socks-to-rtc.ts:292]: Done.

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 8 above)
Comment (bemasc) [inline src/socks-to-rtc/socks-to-rtc.ts]: Done.

File: src/rtc-to-net/rtc-to-net.ts (hunk 9 above)
Comment (bemasc) [inline src/rtc-to-net/rtc-to-net.ts]: Done.

**Comment by bemasc** [comment]
I think I've hit all your comments.

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 5 above)
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts:292]: :+1:

File: src/socks-to-rtc/socks-to-rtc.spec.ts (hunk 6 above)
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.spec.ts:110]: :+1:

File: src/rtc-to-net/rtc-to-net.ts (hunk 7 above)
Comment (trevj) [inline src/rtc-to-net/rtc-to-net.ts]: :+1:

File: src/socks-to-rtc/socks-to-rtc.ts (hunk 8 above)
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: :+1:

File: src/rtc-to-net/rtc-to-net.ts (hunk 9 above)
Comment (trevj) [inline src/rtc-to-net/rtc-to-net.ts]: :+1:

File: src/churn/churn.ts (hunk 10 above)
Comment (trevj) [inline src/churn/churn.ts:24]: Agreed.

File: src/churn/churn.ts [hunk 15]
```diff
@@ -244,138 +303,64 @@ module Churn {
...
     }
 
-    public onceConnected = () : Promise<WebRtc.ConnectionAddresses> => {
+    public onceConnected = this.onceChurnSetup_.then(() => {
       // obfuscatedConnection_ doesn't exist until onceChurnSetup_ fulfills.
-      return this.onceChurnSetup_.then(() => {
-        return this.obfuscatedConnection_.onceConnected();
-      });
-    }
+      return this.obfuscatedConnection_.onceConnected
```
Comment (trevj) [inline src/churn/churn.ts:341]: missing semicolon

File: src/churn/churn.ts (hunk 12 above)
Comment (trevj) [inline src/churn/churn.ts]: Nice. :+1:

File: src/churn/churn.ts (hunk 13 above)
Comment (trevj) [inline src/churn/churn.ts]: :+1:

**Comment by trevj** [comment]
:100:

File: src/churn/churn.ts (hunk 15 above)
Comment (bemasc) [inline src/churn/churn.ts:341]: Fixed.

**Pull request merged**